runtime/vm/OutOfLineINL_jdk_internal_foreign_abi_DowncallLinker.cpp

```diff
@@ -0,0 +1,211 @@
+ * forth in the Eclipse Public License, v. 2.0 are satisfied: GNU
+ * General Public License, version 2 with the GNU Classpath
+ * Exception [1] and GNU General Public License, version 2 with the
+ * OpenJDK Assembly Exception [2].
+ *
+ * [1] https://www.gnu.org/software/classpath/license.html
+ * [2] http://openjdk.java.net/legal/assembly-exception.html
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-exception-2.0 OR LicenseRef-GPL-2.0 WITH Assembly-exception
+ *******************************************************************************/
+
+#include "VMHelpers.hpp"
+#include "BytecodeAction.hpp"
+#include "UnsafeAPI.hpp"
+#include "j9vmnls.h"
+#include "OutOfLineINL.hpp"
+#include "LayoutFFITypeHelpers.hpp"
+#include "AtomicSupport.hpp"
+
+extern "C" {
+
+#if JAVA_SPEC_VERSION >= 19
+/* jdk.internal.foreign.abi.DowncallLinker: private static synchronized native void resolveRequiredFields(); */
+VM_BytecodeAction
+OutOfLineINL_jdk_internal_foreign_abi_DowncallLinker_resolveRequiredFields(J9VMThread *currentThread, J9Method *method)
+{
+	VM_BytecodeAction rc = EXECUTE_BYTECODE;
+	J9JavaVM *vm = currentThread->javaVM;
+	J9ConstantPool *jclConstantPool = (J9ConstantPool *)vm->jclConstantPool;
+	const int cpEntryNum = 2;
+	U_16 cpIndex[cpEntryNum] = {
+				J9VMCONSTANTPOOL_JDKINTERNALFOREIGNABIDOWNCALLLINKER_CIFNATIVETHUNKADDR,
+				J9VMCONSTANTPOOL_JDKINTERNALFOREIGNABIDOWNCALLLINKER_ARGTYPESADDR
+			};
+
+	VM_OutOfLineINL_Helpers::buildInternalNativeStackFrame(currentThread, method);
+	for (int i = 0; i < cpEntryNum; i++) {
+		J9RAMFieldRef *cpFieldRef = ((J9RAMFieldRef*)jclConstantPool) + cpIndex[i];
+		UDATA const flags = cpFieldRef->flags;
+		UDATA const valueOffset = cpFieldRef->valueOffset;
+
+		if (!VM_VMHelpers::instanceFieldRefIsResolved(flags, valueOffset)) {
+			resolveInstanceFieldRef(currentThread, NULL, jclConstantPool, cpIndex[i], J9_RESOLVE_FLAG_NO_THROW_ON_FAIL | J9_RESOLVE_FLAG_JCL_CONSTANT_POOL, NULL);
+			if (VM_VMHelpers::exceptionPending(currentThread)) {
+				rc = GOTO_THROW_CURRENT_EXCEPTION;
+				goto done;
+			}
+		}
+	}
+	VM_OutOfLineINL_Helpers::restoreInternalNativeStackFrame(currentThread);
+
+done:
+	VM_OutOfLineINL_Helpers::returnVoid(currentThread, 0);
+	return rc;
+}
+
+/**
+ * jdk.internal.foreign.abi.DowncallLinker: private native void initCifNativeThunkData(String[] argLayouts, String retLayout, boolean newArgTypes);
+ *
+ * @brief Prepare the prep_cif for the native function specified by the arguments/return layouts
+ * @param argLayouts[in] A c string array describing the argument layouts
+ * @param retLayout[in] A c string describing the return layouts
+ * @param newArgTypes[in] a flag determining whether to create a new ffi_type array for arguments
+ * @return void
+ */
+VM_BytecodeAction
+OutOfLineINL_jdk_internal_foreign_abi_DowncallLinker_initCifNativeThunkData(J9VMThread *currentThread, J9Method *method)
```

Comment:
Codewise it is pretty much the same for Java17/18(`ProgrammableInvoker`) and 19(`DowncallLinker`) but this can be merged into one line if we are able to wrap them up to an internal package (e.g. `openj9.internal.foreign`) to share the same core functionalities between Java17/18 and 19 (under investigation).

Reply:
The two files are removed and replaced with `/runtime/vm/OutOfLineINL_openj9_internal_foreign_abi_InternalDowncallHandler.cpp` based on our latest solution to avoid duplicating code in Java17/18 and Java19